File: Hard/TrappingRainWater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) == 0:
            return 0
        
        left = 0
        right = len(height) - 1
        
        leftMax = height[left]
        rightMax = height[right]
        
        ans = 0
        
        while left < right:
            if height[left] < height[right]:
                if height[left] < leftMax:
                    ans += leftMax - height[left]
                else:
                    leftMax = height[left]
                left += 1
                
            else:
                if height[right] < rightMax:
                    ans += rightMax - height[right]
                else:
                    rightMax = height[right]
                right -= 1
                
        return ans

File: Hard/TrappingRainWater.java
/*
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.

For example, 
Given [0,1,0,2,1,0,1,3,2,1,2,1], return 6.
*/

public class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int res = 0, maxLeft = 0, maxRight = 0;
        while (left <= right){
            if (height[left] <= height[right]){
                if (height[left] > maxLeft) maxLeft = height[left];
                else res += (maxLeft - height[left]);
                left++;
            }
            else {
                if (height[right] > maxRight) maxRight = height[right];
                else res += maxRight - height[right];
                right--;
            }
        }
        return res;
    }
}